src/dotty/tools/dotc/transform/IsInstanceOfEvaluator.scala

```diff
@@ -0,0 +1,166 @@
+
+    /** Rewrites the select to a boolean if `to` is false or if the qualifier
+     *  is a value class.
+     *
+     *  If `to` is set to true and the qualifier is not a primitive, the
+     *  instanceOf is replaced by a null check, since:
+     *
+     *  `scrutinee.isInstanceOf[Selector]` if `scrutinee eq null`
+     */
+    def rewrite(tree: Select, to: Boolean): Tree =
+      if (!to || !tree.qualifier.tpe.widen.derivesFrom(defn.AnyRefAlias))
+        Literal(Constant(to))
+      else
+        Apply(tree.qualifier.select(defn.Object_ne), List(Literal(Constant(null))))
+
+    /** Attempts to rewrite TypeApply to either `scrutinee ne null` or a
+     *  constant
+     */
+    def evalTypeApply(tree: TypeApply): Tree =
+      if (tree.symbol != defn.Any_isInstanceOf) tree
+      else tree.fun match {
+        case s: Select => {
+          val scrutinee = erasure(s.qualifier.tpe.widen)
+          val selector  = erasure(tree.args.head.tpe.widen)
+
+          val scTrait = scrutinee.typeSymbol is Trait
+          val scClass =
+            scrutinee.typeSymbol.isClass &&
+            !(scrutinee.typeSymbol is Trait) &&
+            !(scrutinee.typeSymbol is Module)
+
+          val scClassNonFinal = scClass && !(scrutinee.typeSymbol is Final)
+          val scFinalClass    = scClass && (scrutinee.typeSymbol is Final)
+
+          val selTrait = selector.typeSymbol is Trait
+          val selClass =
+            selector.typeSymbol.isClass &&
+            !(selector.typeSymbol is Trait) &&
+            !(selector.typeSymbol is Module)
+
+          val selClassNonFinal = scClass && !(selector.typeSymbol is Final)
+          val selFinalClass    = scClass && (selector.typeSymbol is Final)
+
+          // Cases ---------------------------------
+          val valueClassesOrAny =
+            ValueClasses.isDerivedValueClass(scrutinee.typeSymbol) ||
+            ValueClasses.isDerivedValueClass(selector.typeSymbol)  ||
+            scrutinee == defn.ObjectType
+
+          val knownStatically = scFinalClass
+
+          val falseIfUnrelated =
+            (scClassNonFinal && selClassNonFinal) ||
+            (scClassNonFinal && selFinalClass)    ||
+            (scTrait && selFinalClass)
+
+          val happens =
+            (scClassNonFinal && selClassNonFinal) ||
+            (scTrait && selClassNonFinal)         ||
+            (scTrait && selTrait)
+
+          val inMatch = s.qualifier.symbol is Case
+
+          if (valueClassesOrAny) tree
+          else if (knownStatically)
+            handleStaticallyKnown(s, scrutinee, selector, inMatch, tree.pos)
+          else if (falseIfUnrelated && scrutinee <:< selector)
+            // scrutinee is a subtype of the selector, safe to rewrite
+            rewrite(s, to = true)
+          else if (falseIfUnrelated && !(selector <:< scrutinee))
+            // selector and scrutinee are unrelated
+            handleFalseUnrelated(s, scrutinee, selector, inMatch)
+          else if (happens) tree
```

Comment:
Agreed, it was kept to showcase all cases - but I'll change it to and note this in a comment.